packages/sw-background-sync-queue/src/lib/queue.js

```diff
@@ -0,0 +1,143 @@
+import stringHash from 'string-hash';
+import idbQHelper from './idbHelper';
+import requestManager from './requestManager';
+import bcmanager from './broadcastManager';
+
+let _queue;
+let _config;
+
+/**
+ * Core queue class that handles all the enqueue and dequeue
+ * as well as cleanup code for the background sync queue
+ * @class Queue
+ */
+class Queue {
+	/**
+	 * Creates an instance of Queue.
+	 *
+	 * @memberOf Queue
+	 */
+	constructor() {
+		this.initialize();
+	}
+
+	/**
+	 * initialize the queue object from the idb
+	 * giving access to any penging queues
+	 * @param {Object} config config that gives values like
+	 * maxAge of requests in iDB
+	 *
+	 * @memberOf Queue
+	 */
+	async initialize(config) {
+		_config = config;
+		_queue = await idbQHelper.get('queue') || [];
+	}
+
+	/**
+	 * push any request to background sync queue which would be played later
+	 * prefferably when network comes back
+	 *
+	 * @param {Request} request request object to be queued by this
+	 * @param {Object} config optional config to override config params
+	 *
+	 * @memberOf Queue
+	 */
+	async push(request, config) {
+		let localConfig = Object.assign({}, _config, config);
+		let hash = stringHash(request.url + JSON.stringify(request.headers)
+			+ await request.text() + Date.now());
+		let queuableRequest =
+			await requestManager.getQueueableRequest(request, localConfig);
+		try{
+			_queue.push(hash);
+
+			// add to queue
+			idbQHelper.put('queue', _queue);
+			idbQHelper.put(hash, queuableRequest);
+
+			// register sync
+			self.registration.sync.register('bgqueue');
+
+			// broadcast the success
+			bcmanager.postMessage({
+				status: 'success',
+				id: hash,
+				request: queuableRequest,
+			});
+		} catch(e) {
+			// broadcast the success
+			bcmanager.postMessage({
+				status: 'failed',
+				request: request,
+			});
+		}
+	}
+
+	/**
+	 * clean up the queue, deleting all the tasks who are either damaged or
```

Comment:
Could we document what results in a task being damaged? It's pretty clear what leads to a maxAge being expired but it's unclear how a task might get into this other state nor whether there are things that could be done to recover from this state (other than be deleted).